File: java/JumpSearch.java
// time complexity; O(sqrt(n))
// space complexity; O(1)

public class JumpSearch
  {
    public static int jumpSearch(int arr[], int x)
    {
      int n = arr.length;
      // find block size to be jumped
      int step = (int)Math.floor(Math.sqrt(n));
      // find block where element may be present
      int prev = 0;
      while (arr[Math.min(step, n)-1] < x)
        {
          prev = step;
          step += (int)Math.floor(Math.sqrt(n));
          if (prev >= n)
            return -1;
        }
      // do linear search in block found above
      while (arr[prev] < x)
        {
          prev++;
          if(prev == Math.min(step, n))
            return -1;
        }
      // if element is found
      if (arr[prev] == x)
        return prev;
      return -1;
    }
  }

File: python/jump_search.py
# time complexity: O(sqrt(N))
# space complexity: O(1)

def jump_search(arr, x):
	n = len(arr)
	step = int(math.sqrt(n))

	# find block where elmeent may be present
	prev = 0
	while arr[min(step, n) - 1] < x:
		prev = step
		step += int(math.sqrt(n))
		if prev >= n:
			return -1

	# do a linear search in the block found above
	while arr[prev] < x:
		prev += 1

	# if we reached next block or end of array, element is not present
	if prev == min(step, n) and arr[prev] != x:
		return -1

	# if element is found
	if arr[prev] == x:
		return prev

return -1 
